README.md
# ogms_releases


Changelog.md
#Changes made between successive versions of OGMS

# OGMSv0.9-v0.11 Changelog #
  * Update to BFO2 ids
  * Modified definition of 'clinical picture' to include dispositions and allow for partial clinical pictures.
  * Filled in placeholder definition of 'pain' with that of [(Smith, Ceusters, Goldberg and Ohrbach, 2011)](http://www.referent-tracking.com/RTU/sendfile/?file=painTokyo1_27_2011.pdf)
  * [Issue 69](http://code.google.com/p/ogms/issues/detail?id=69) Modified definition of 'diagnostic process' to allow the diagnostic statement (output) to be about a disease, disorder, or syndrome, or that a patient has none of these.
  * [Issue 72](http://code.google.com/p/ogms/issues/detail?id=72) Changed 'prognosis' definition to refer to some future part of a disease course.
  * [Issue 73](http://code.google.com/p/ogms/issues/detail?id=73) Added terms 'relapse' and 'remission'.

# OGMSv0.8-v0.9 Changelog #
  * [Issue 60](http://code.google.com/p/ogms/issues/detail?id=60) closed, added 'diagnostic process', modified term 'diagnosis'.
  * [Issue 61](http://code.google.com/p/ogms/issues/detail?id=61) closed, added term 'injury'.
  * [Issue 65](http://code.google.com/p/ogms/issues/detail?id=65) closed, superseded by [Issue 66](https://code.google.com/p/ogms/issues/detail?id=66).
  * [Issue 66](http://code.google.com/p/ogms/issues/detail?id=66) left open for refinement of 'treatment', added term 'prophylaxis'

# OGMSv0.7-v0.8 Changelog #
  * [Issue 54](http://code.google.com/p/ogms/issues/detail?id=54) closed, added 'health care process', 'health care encounter', and stubs for subtypes.  See also ogms-discuss thread on the issue [here](http://groups.google.com/group/ogms-discuss/browse_thread/thread/a2dbc2ed1dff99d6/bf0e22e8a344c173#bf0e22e8a344c173).

# OGMSv0.6-v0.7 Changelog #
  * Unobsoleted defined class (BFO:0000076) to accommodate signs and symptoms which could be either continuants or occurrents (since these are disjoint per bfo, any classifier would complain if they are asserted as quality OR processual entity).
  * [Issue 12](http://code.google.com/p/ogms/issues/detail?id=12) closed, changed definition of symptom per discussion.
  * Changed definition of sign per discussion.
  * [Issue 36](http://code.google.com/p/ogms/issues/detail?id=36) closed without action, CPR-OGMS overlaps will be handled by mutual referencing and/or mireot where necessary.
  * [Issue 49](http://code.google.com/p/ogms/issues/detail?id=49) closed, fixed misspelling.
  * [Issue 50](http://code.google.com/p/ogms/issues/detail?id=50) closed, reclassified suggested processes as processual entities in preparation for BFO 2.0

# OGMSv0.5-v0.6 Changelog #
  * [Issue 32](http://code.google.com/p/ogms/issues/detail?id=32) closed, changed definition of 'syndrome', reclassified as bfo:quality.
  * Created term 'acute disease course' (OGMS:0000094)
  * Obsoleted 'bodily feature of a patient'. Reclassified its children as bfo:quality and modified definitions accordingly.  Note, per previous discussions, this makes 'sign' and 'symptom' into qualities (blocks all processual entities from being sings or symptoms and requires that they are exhibited when borne).  More discussion will surely follow.
  * [Issue 41](http://code.google.com/p/ogms/issues/detail?id=41) left open, obsoleted 'defined\_class' (since 'bodily feature of a patient' was its final child)
  * Created term 'inflamation process' (OGMS:0000095)

# OGMSv0.4-v0.5 Changelog #
  * [Issue 28](http://code.google.com/p/ogms/issues/detail?id=28) closed, added 'congenital malformation' (OGMS:0000089).
  * [Issue 35](http://code.google.com/p/ogms/issues/detail?id=35) left open, added 'treatment' (OGMS:0000090) and 'convalescence' (OGMS:0000091).
  * [Issue 38](http://code.google.com/p/ogms/issues/detail?id=38) closed, added 'life course' (OGMS:0000092)
  * [Issue 44](http://code.google.com/p/ogms/issues/detail?id=44) closed, determined that 'state' does not belong in OGMS.

# OGMSv0.3-v0.4 Changelog #
Changes made for OGMSv0.4
  * OBO file: Fixed typo 'occurrentno' (BFO:0000002) to 'occurrent'.
  * OBO file: Fixed typo 'preclinical manifestationof a disease' (OGMS:0000026) to 'preclinical manifestation of a disease'
  * [Issue 3](http://code.google.com/p/ogms/issues/detail?id=3) left open, changed definition of 'disorder' (OGMS:0000045), changed it to 'material entity', added term 'extended organism' (OGMS:0000087), and corrected 'borne' to 'part of' in the definition of 'constitutional genetic disorder' to reflect the fact that disorders are independent continuants.
  * [Issue 12](http://code.google.com/p/ogms/issues/detail?id=12) left open, added term 'patient symptom report' (OGMS:0000088)
  * [Issue 20](http://code.google.com/p/ogms/issues/detail?id=20) closed, changed 'disease course' (OGMS:0000063) from 'process aggregate' to 'processual entity'
  * [Issue 21](http://code.google.com/p/ogms/issues/detail?id=21) closed, changed 'clinical history taking' (OGMS:0000055), 'laboratory test' (OGMS:0000056), and 'physical examination' (OGMS:0000057) from 'fiat process part' to 'processual entity'

# OGMSv0.2-v0.3 Changelog #
Changes made for OGMSv0.3
  * OBO file: Added dbxrefs to the issue tracker for traceability
  * OBO and OWL files: Added `obsolete_` prefix to obsoleted terms
  * [Issue 4](http://code.google.com/p/ogms/issues/detail?id=4) closed without change
  * [Issue 5](http://code.google.com/p/ogms/issues/detail?id=5) closed without change
  * [Issue 6](http://code.google.com/p/ogms/issues/detail?id=6) closed opened [issue 37](https://code.google.com/p/ogms/issues/detail?id=37) to clairfy
  * [Issue 8](http://code.google.com/p/ogms/issues/detail?id=8) closed, modified 'preclinical manifestation' (OGMS:0000026)
  * [Issue 9](http://code.google.com/p/ogms/issues/detail?id=9) closed without change
  * [Issue 10](http://code.google.com/p/ogms/issues/detail?id=10) closed without change

trunk/src/tools/ogms2owl2.lisp
;; another obo to owl converter

;; (setq obo (make-instance 'obo :path "~/repos/ogms/trunk/src/ontology/ogms.obo"))
;; (read-obo obo)
;; (ogms-to-owl obo) -> ~/Desktop/ogms.owl

(defun getf-multiple (plist indicator)
  (loop for (key value) on plist by #'cddr when (eq indicator key) collect value))

(defun ogms-to-owl (obo &optional (dest "ogms:ontology;ogms.owl"))
  (flet ((uri-ify (id)
	   (cond ((equal id "BFO:0000076")
		  !bfo:Entity)
		 ((#"matches" id "^BFO:.*")
		  (make-uri (second (assoc id *bfo-map* :test 'equalp))))
		 ((#"matches" id "^OGMS:.*")
		  (make-uri (#"replaceFirst" id "OGMS:" "http://purl.obolibrary.org/obo/OGMS_")))
		 ((#"matches" id "^OBO_REL:.*")
		  (make-uri (#"replaceFirst" id "^OBO_REL:" "http://www.obofoundry.org/ro/ro.owl#")))
		 ((#"matches" id "^IAO:.*")
		  (make-uri (#"replaceFirst" id "^IAO:" "http://purl.obolibrary.org/obo/IAO_")))
		 (t (error "don't know how to convert id ~a to uri" id)))))
    (with-ontology ogms (:about "http://purl.obolibrary.org/obo/ogms.owl" :base "http://purl.obolibrary.org/obo/ogms.owl" :collecting t)
	((asq 
	  (imports !<http://www.ifomis.org/bfo/1.1>)
	 (imports !<http://purl.obolibrary.org/obo/iao/ontology-metadata.owl>)
	 (imports !<http://protege.stanford.edu/plugins/owl/dc/protege-dc.owl>))
	 (let ((ogms !<http://purl.obolibrary.org/obo/ogms.owl>))
	   (as
	    `(annotation-assertion !rdfs:comment  ,ogms "The Ontology for General Medical Science (OGMS) is based on the papers Toward an Ontological Treatment of Disease and Diagnosis and On Carcinomas and Other Pathological Entities. The ontology attempts to address some of the issues raised at the Workshop on Ontology of Diseases (Dallas, TX). OGMS was formerly called the clinical phenotype ontology. Terms from OGMS hang from the Basic Formal Ontology.

The latest version of OGMS is always available at http://purl.obolibrary.org/obo/ogms.owl
This is version 2009-12-03 aka '0.3'
http://purl.obolibrary.org/obo/ogms/2009-12-09/ogms.owl

The OGMS developer site is http://code.google.com/p/ogms/
The discussion group is http://groups.google.com/group/ogms-discuss

If you are interested in participating in the development of OGMS, please send email to albertgoldfain@gmail.com. Be sure to include a google-account username with your request (this will be the username associated with a gmail address).")
	    `(annotation-assertion  !rdfs:comment  ,ogms "This ontology is in early development. Expect it to change.")
	    `(annotation-assertion  !dc:creator  ,ogms "Richard Scheuermann")
	    `(annotation-assertion  !dc:creator  ,ogms "Barry Smith")
	    `(annotation-assertion  !dc:creator  ,ogms "Werner Ceusters")
	    `(annotation-assertion  !dc:creator  ,ogms "Albert Goldfain")
	    `(annotation-assertion  !dc:creator  ,ogms "Lindsay Cowell")
	    `(annotation-assertion  !dc:creator  ,ogms "Anand Kumar")
	    `(annotation-assertion  !dc:creator  ,ogms "Cornelius Rosse")
	    `(annotation-assertion  !dc:creator  ,ogms "Bill Hogan")
	    `(annotation-assertion  !dc:creator  ,ogms "Sivaram Arabandi")
	    `(annotation-assertion  !dc:creator  ,ogms "Bill Hogan")
	    `(annotation-assertion  !dc:creator  ,ogms "Daniel Merico")
	    `(annotation-assertion  !dc:creator  ,ogms "Alan Ruttenberg"))
	   (asq
	    (declaration (annotation-property !obo:IAO_0000115))
	    (annotation-assertion !rdfs:label !obo:IAO_0000115 "definition")
	    (declaration (annotation-property !obo:IAO_0000117))
	    (annotation-assertion  !rdfs:label !obo:IAO_0000117 "definition editor")
	    (declaration (annotation-property !obo:IAO_0000119))
	    (annotation-assertion  !rdfs:label !obo:IAO_0000119 "definition source")
	    (declaration (annotation-property !obo:IAO_0000116))
	    (annotation-assertion  !rdfs:label !obo:IAO_0000116 "editor note")
	    (declaration (annotation-property !obo:IAO_0000232))
	    (annotation-assertion  !rdfs:label !obo:IAO_0000232 "curator note")
	    (declaration (annotation-property !dc:date)))
	   (as `(annotation-assertion !dc:date ,ogms "2009-08-07")))
	  (loop for record in (terms obo) 
	    for name = (getf (cdr record) :name)
	    for id = (getf (cdr record) :id)
	    for is_a = (getf-multiple (cdr record) :is_a)
	    for xrefs = (getf-multiple (cdr record) :xref)
	    for def = (getf (cdr record) :def)
	    for created-by = (getf (cdr record) :created_by)
	    for creation-date = (getf (cdr record) :creation_date)
	     for uri = (uri-ify id)
	    if (equal name "undefined primitive terms") do (setq name "_undefined primitive term")
	    unless (or (equal id !owl:Thing) (equal id !bfo:Entity) (#"matches" id "BFO.*") (equal id "OGMS:0000072"))
	     do
	       (as `(declaration (class ,uri)))
	       (unless (#"matches" id "IAO.*") (as `(annotation-assertion  !rdfs:label ,uri ,name)))
	       (if (car def) (as `(annotation-assertion !obo:IAO_0000115 ,uri  ,(format nil "~a@en" (car def)))))
	       (if (third def) (as `(annotation-assertion !obo:IAO_0000119 ,uri  ,(third def))))
	       (if created-by (as `(annotation-assertion !obi:IAO_0000117 ,uri  ,(#"replaceAll" created-by "agoldfain" "Albert Goldfain"))))
	       (if creation-date (as `(annotation-assertion !obo:IAO_0000232 ,uri  ,(format nil "creation date: ~a" creation-date))))
	       (if (member "BFO:0000076" is_a :test 'equalp)
		   (as `(annotation-assertion !obi:IAO_0000116 ,uri  "note: defined class")))
	       (if (car xrefs)
		   (loop for xref in xrefs do
			(as `(annotation-assertion !oboinowl:hasDbXref ,uri  ,(format nil "~a" xref)))))
	       (if (getf (cdr record) :is_obsolete)
		   (as `(subclassof ,uri !oboinowl:ObsoleteClass))
		   (loop for is in is_a do (as `(subclassof ,uri ,(uri-ify is)))))))
      (write-rdfxml ogms dest))))

(defun ogms-convert (&optional (source "ogms:ontology;ogms.obo") (dest "ogms:ontology;ogms.owl"))
  (let ((obo (make-instance 'obo :path source)))
    (read-obo obo)
    (ogms-to-owl obo dest)))

(defparameter *bfo-map* 
  '(("BFO:0000000" "http://www.ifomis.org/bfo/1.1#Entity")
   ("BFO:0000001" "http://www.ifomis.org/bfo/1.1/snap#Continuant")
   ("BFO:0000003" "http://www.ifomis.org/bfo/1.1/snap#DependentContinuant")
   ("BFO:0000004" "http://www.ifomis.org/bfo/1.1/snap#IndependentContinuant")
   ("BFO:0000005" "http://www.ifomis.org/bfo/1.1/snap#GenericallyDependentContinuant")
   ("BFO:0000006" "http://www.ifomis.org/bfo/1.1/snap#SpecificallyDependentContinuant")
   ("BFO:0000007" "http://www.ifomis.org/bfo/1.1/snap#Quality")
   ("BFO:0000008" "http://www.ifomis.org/bfo/1.1/snap#RealizableEntity")
   ("BFO:0000009" "http://www.ifomis.org/bfo/1.1/snap#Disposition")
   ("BFO:0000010" "http://www.ifomis.org/bfo/1.1/snap#Role")
   ("BFO:0000011" "http://www.ifomis.org/bfo/1.1/snap#Function")
   ("BFO:0000012" "http://www.ifomis.org/bfo/1.1/snap#Capability")
   ("BFO:0000041" "http://www.ifomis.org/bfo/1.1/snap#MaterialEntity")
   ("BFO:0000042" "http://www.ifomis.org/bfo/1.1/snap#FiatObjectPart")
   ("BFO:0000043" "http://www.ifomis.org/bfo/1.1/snap#Object")
   ("BFO:0000044" "http://www.ifomis.org/bfo/1.1/snap#ObjectAggregate")
   ("BFO:0000002" "http://www.ifomis.org/bfo/1.1/span#Occurrent")
   ("BFO:0000052" "http://www.ifomis.org/bfo/1.1/span#ProcessualEntity")
   ("BFO:0000053" "http://www.ifomis.org/bfo/1.1/span#FiatProcessPart")
   ("BFO:0000058" "http://www.ifomis.org/bfo/1.1/span#Process")
   ("BFO:0000062" "http://www.ifomis.org/bfo/1.1/span#ProcessAggregate")))